hoomd/BoxResizeUpdaterGPU.cu

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#include "BoxResizeUpdaterGPU.cuh"
+
+namespace hoomd
+    {
+namespace kernel
+    {
+
+__global__ void gpu_box_resize_updater_kernel(Scalar4* d_pos,
+                                              const BoxDim cur_box,
+                                              const BoxDim new_box,
+                                              const unsigned int* d_group_members,
+                                              const unsigned int group_size)
+    {
+    int group_idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if (group_idx < group_size)
+        {
+        unsigned int idx = d_group_members[group_idx];
+
+        Scalar4 pos = d_pos[idx];
+
+        Scalar3 fractional_pos = cur_box.makeFraction(make_scalar3(pos.x, pos.y, pos.z));
+
+        Scalar3 scaled_pos = new_box.makeCoordinates(fractional_pos);
+        d_pos[idx].x = scaled_pos.x;
+        d_pos[idx].y = scaled_pos.y;
+        d_pos[idx].z = scaled_pos.z;
+        }
+    }
+
+__global__ void
+gpu_box_wrap_kernel(unsigned int N, Scalar4* d_pos, int3* d_image, const BoxDim new_box)
+    {
+    int idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if (idx < N)
+        {
+        new_box.wrap(d_pos[idx], d_image[idx]);
+        }
+    }
+
+hipError_t gpu_box_resize_updater(const unsigned int N,
+                                  Scalar4* d_pos,
+                                  const BoxDim& cur_box,
+                                  const BoxDim& new_box,
+                                  const unsigned int* d_group_members,
+                                  const unsigned int group_size,
+                                  int3* d_image)
+    {
+    int block_size = 256;
```

Comment:
Is it recommended for all new code to use the `GPUPartition` to split up work? Or is this only used in beefy kernels like the integrator?

Reply:
Also, do you recommend keeping or omitting the `hipDeviceSynchronize()` call at the end. I added it in caution, and to return the error state, but it will block all kernels?

Reply:
I would skip doing GPU partitions. This is a code path that a previous developer added to some kernels to increase performance on the dense OLCF Summit nodes with NVLINK interconnects. I'm not aware of anyone actively using this code path any longer and I don't personally use or test it. Summit will be decommissioned at the end of the year and very few other HPC systems are built with high speed GPU<->GPU interconnects. 

Don't call `hipDeviceSynchronize() or check for error codes` unless `isCUDAErrorCheckingEnabled()` is true. Kernel launch latency and synchronization is expensive and the only way to keep performance high is to launch as many kernels asynchronously as possible. `CHECK_CUDA_ERROR` will call `hipDeviceSynchronize` to check for kernel launch errors.